tests/e2e/tests/istioctl/istioctl_test.go

```diff
@@ -0,0 +1,355 @@
+}
+
+func (t *testConfig) Teardown() error {
+	return nil
+}
+
+func waitForRule(ruleKey string, retries int, durBetweenAttempts time.Duration) error {
+
+	for retry := 0; retry < retries; retry++ {
+		time.Sleep(durBetweenAttempts)
+		var output string
+		var err error
+		if output, err = tc.Kube.Istioctl.RunWithOutput([]string{"get", "route-rules"},
+			"", "", false); err != nil {
+			return err
+		}
+		if strings.Contains(output, ruleKey) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("created rule did not exist after %v", time.Duration(retries)*durBetweenAttempts)
+}
+
+func waitForRuleRemoval(ruleKey string, retries int, durBetweenAttempts time.Duration) error {
+
+	for retry := 0; retry < retries; retry++ {
+		time.Sleep(durBetweenAttempts)
+		var output string
+		var err error
+		if output, err = tc.Kube.Istioctl.RunWithOutput([]string{"get", "route-rules"},
+			"", "", false); err != nil {
+			return err
+		}
+		if !strings.Contains(output, ruleKey) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("created rule still exists after %v", time.Duration(retries)*durBetweenAttempts)
+}
+
+// Mixer rules don't have names so we just wait for a rule of the appropriate "kind" (e.g. "quota")
+func waitForMixerRuleType(scope, subject, kind string, retries int, durBetweenAttempts time.Duration) error {
+
+	for retry := 0; retry < retries; retry++ {
+		time.Sleep(durBetweenAttempts)
+		var output string
+		var err error
+		if output, err = tc.Kube.Istioctl.RunWithOutput([]string{"mixer", "rule", "get", scope, subject},
+			"", "", false); err != nil {
+			return err
+		}
+		if strings.Contains(output, kind) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("created mixer rule kind did not exist after %v", time.Duration(retries)*durBetweenAttempts)
+}
+
+func rulePath(ruleFile string) string {
+	return filepath.Join(tc.rulesDir, ruleFile)
+}
+
+// TestInvalidRuleDetected verifies attempting to create invalid rule fails and generates an error message explaining the problem
+func TestInvalidRuleDetected(t *testing.T) {
+	invalidRules := []ruleFileAndCreateResponse{
+		{
+			ruleFile: "route-rule-101.yaml",
+			response: regexp.MustCompile("percent invalid:  must be in range 0..100"),
+		},
+	}
+
+	for _, invalidRule := range invalidRules {
+		output, err := tc.Kube.Istioctl.RunWithOutput([]string{"create", "--file", rulePath(invalidRule.ruleFile)},
+			"", "", true)
+		if err == nil {
+			glog.Error(fmt.Sprintf("Expected failure but got success %v", output))
+			t.Fatal(fmt.Sprintf("Expected failure but got success %v", output))
+		}
+		if !invalidRule.response.Match([]byte(output)) {
+			t.Errorf("create file %v did not match %v.  Output: %v", invalidRule.ruleFile, invalidRule.response, output)
+		}
+	}
+}
+
+// TestCRUD tests Create, Read, Update, Delete operations for a rule by performing them and scraping the output.
+// Note: If you are improving the output of istioctl, you will have to modify the regexps here used to detect expected output.
```

Comment:
You have a good point.

I would have rather put this code in istio/manager for that very reason.  I would still be happy to move the code there if someone can suggest the correct place to hook it into the test: after istioctl has been compiled, and with the possibility of standing up a manager and mixer for it to talk to.

Ideally this would be a "promotion" test rather than something that runs on every build, so that we can test a real client and server together before releasing it.

I asked a few questions about the presubmit (which is in an integration directory), the manager smoketest, and the istio/istio tests.  I didn't really got answers so decided to write the code bring the discussion here.  Obviously you can force-merge one and then do the other, or disable a test here and re-enable it.  Both methods are a pain.